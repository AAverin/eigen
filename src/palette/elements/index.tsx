export * from "./Cards"
export * from "./Touchable"
export * from "./Typography"
