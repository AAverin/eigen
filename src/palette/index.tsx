export * from "./atoms"
export * from "./organisms"
export * from "./elements"
export * from "./Theme"
