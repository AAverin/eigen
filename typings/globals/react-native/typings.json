{
  "resolution": "main",
  "tree": {
    "src": "https://raw.githubusercontent.com/DefinitelyTyped/DefinitelyTyped/4c7d03b088d4e71cd9fe9c66ab7bd363049077c3/react-native/react-native.d.ts",
    "raw": "registry:dt/react-native#0.29.0+20160719232534",
    "typings": "https://raw.githubusercontent.com/DefinitelyTyped/DefinitelyTyped/4c7d03b088d4e71cd9fe9c66ab7bd363049077c3/react-native/react-native.d.ts"
  }
}
