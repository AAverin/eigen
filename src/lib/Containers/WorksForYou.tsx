import { Box, Flex, Theme } from "@artsy/palette"
import React from "react"
import {
  ListView,
  ListViewDataSource,
  NativeModules,
  RefreshControl,
  ScrollView,
  StyleSheet,
  TextStyle,
  View,
  ViewStyle,
} from "react-native"
import { createPaginationContainer, graphql, RelayPaginationProp } from "react-relay"

import Events from "../NativeModules/Events"

import { PAGE_SIZE } from "lib/data/constants"

import { WorksForYou_query } from "__generated__/WorksForYou_query.graphql"
import Spinner from "lib/Components/Spinner"
import ZeroState from "lib/Components/States/ZeroState"
import Notification from "lib/Components/WorksForYou/Notification"
import colors from "lib/data/colors"
import { isCloseToBottom } from "lib/utils/isCloseToBottom"

interface Props {
  relay: RelayPaginationProp
  query: WorksForYou_query
}

interface State {
  dataSource: ListViewDataSource | null
  isRefreshing: boolean
  loadingContent: boolean
}

export class WorksForYou extends React.Component<Props, State> {
  // TODO: This `| any` is a hack workaround to a typing bug in https://github.com/artsy/emission/pull/504/
  scrollView?: ScrollView | any
  currentScrollOffset?: number = 0

  constructor(props: Props) {
    super(props)

    const notifications: object[] = this.props.query.me.followsAndSaves.notifications.edges.map(edge => edge.node)
    if (this.props.query.selectedArtist) {
      notifications.unshift(this.formattedSpecialNotification())
    }

    const dataSource =
      notifications.length &&
      new ListView.DataSource({ rowHasChanged: (row1, row2) => row1 !== row2 }).cloneWithRows(notifications)

    this.state = {
      dataSource,
      isRefreshing: false,
      loadingContent: false,
    }
  }

  componentDidMount() {
    // Update read status in gravity
    NativeModules.ARTemporaryAPIModule.markNotificationsRead(error => {
      if (error) {
        console.warn(error)
      } else {
        Events.postEvent({
          name: "Notifications read",
          source_screen: "works for you page",
        })
      }
    })
  }

  formattedSpecialNotification() {
    const artist = this.props.query.selectedArtist

    return {
      // This is just some unique ID, don’t rely on MP being able to retrieve a notification by this ID.
      id: `notification-${artist.slug}`,
      message: artist.artworks.edges.length + (artist.artworks.edges.length > 1 ? " Works Added" : " Work Added"),
      artists: artist.name,
      artworks: artist.artworks,
      image: {
        resized: {
          url: artist.image.resized.url,
        },
      },
      artistHref: artist.href,
    }
  }

  fetchNextPage = () => {
    if (!this.props.relay.hasMore() || this.props.relay.isLoading()) {
      return
    }
    this.setState({ loadingContent: true }, () => {
      this.props.relay.loadMore(PAGE_SIZE, error => {
        if (error) {
          // FIXME: Handle error
          console.error("WorksForYou.tsx", error.message)
        }

        const notifications: object[] = this.props.query.me.followsAndSaves.notifications.edges.map(edge => edge.node)

        // Make sure we maintain the special notification if it exists
        if (this.props.query.selectedArtist) {
          notifications.unshift(this.formattedSpecialNotification())
        }

        this.setState(
          {
            dataSource: this.state.dataSource.cloneWithRows(notifications),
          },
          () => {
            this.setState({ loadingContent: false })
          }
        )
      })
    })
  }

  handleRefresh = () => {
    this.setState({ isRefreshing: true })
    this.props.relay.refetchConnection(PAGE_SIZE, error => {
      if (error) {
        // FIXME: Handle error
        console.error("WorksForYou.tsx #handleRefresh", error.message)
      }
      this.setState({ isRefreshing: false })
    })
  }

  render() {
    const hasNotifications = this.state.dataSource

    /* If showing the empty state, the ScrollView should have a {flex: 1} style so it can expand to fit the screen.
       otherwise, it should not use any flex growth.
    */
    return (
      <ScrollView
        contentContainerStyle={hasNotifications ? {} : styles.container}
        onScroll={isCloseToBottom(this.fetchNextPage)}
        scrollEventThrottle={100}
        ref={scrollView => (this.scrollView = scrollView)}
        refreshControl={<RefreshControl refreshing={this.state.isRefreshing} onRefresh={this.handleRefresh} />}
      >
        <View style={{ flex: 1 }}>{hasNotifications ? this.renderNotifications() : this.renderEmptyState()}</View>
      </ScrollView>
    )
  }

  renderNotifications() {
    const { loadingContent } = this.state

    return (
      <Theme>
        <ListView
          dataSource={this.state.dataSource}
          renderRow={data => <Notification notification={data} />}
          renderSeparator={(sectionID, rowID) =>
            <View key={`${sectionID}-${rowID}`} style={styles.separator} /> as React.ReactElement<{}>
          }
          renderFooter={() => (
            <>
              {loadingContent && (
                <Box p={2} style={{ height: 50 }}>
                  <Flex style={{ flex: 1 }} flexDirection="row" justifyContent="center">
                    <Spinner />
                  </Flex>
                </Box>
              )}
            </>
          )}
          scrollEnabled={false}
        />
      </Theme>
    )
  }

  renderEmptyState() {
    return (
      <ZeroState
        title="You haven’t followed any artists yet."
        subtitle="Follow artists to see new works that have been added to Artsy"
      />
    )
  }
}

interface Styles {
  container: ViewStyle
  title: TextStyle
  separator: ViewStyle
}

const styles = StyleSheet.create<Styles>({
  container: {
    flex: 1,
  },
  title: {
    marginTop: 20,
    fontSize: 20,
  },
  separator: {
    height: 1,
    backgroundColor: colors["gray-regular"],
  },
})

const WorksForYouContainer = createPaginationContainer(
  WorksForYou,
  {
    query: graphql`
      fragment WorksForYou_query on Query
        @argumentDefinitions(
          count: { type: "Int", defaultValue: 10 }
          cursor: { type: "String" }
          selectedArtist: { type: "String!", defaultValue: "" }
          sort: { type: "ArtworkSorts" }
        ) {
        me {
          followsAndSaves {
            notifications: bundledArtworksByArtistConnection(sort: PUBLISHED_AT_DESC, first: $count, after: $cursor)
              @connection(key: "WorksForYou_notifications") {
              pageInfo {
                hasNextPage
                endCursor
              }
              edges {
                node {
                  id
                  ...Notification_notification
                }
              }
            }
          }
        }
        selectedArtist: artist(id: $selectedArtist) {
          slug
          href
          name
          image {
            resized(height: 80, width: 80) {
              url
            }
          }
          artworks: artworksConnection(sort: PUBLISHED_AT_DESC, first: 6) {
            edges {
              node {
                ...GenericGrid_artworks
              }
            }
          }
        }
      }
    `,
  },
  {
    direction: "forward",
    getConnectionFromProps(props) {
      return props.query.me.followsAndSaves.notifications
    },
    getFragmentVariables(prevVars, totalCount) {
      return {
        ...prevVars,
        count: totalCount,
      }
    },
    getVariables(_props, { count, cursor }, fragmentVariables) {
      return {
        // in most cases, for variables other than connection filters like
        // `first`, `after`, etc. you may want to use the previous values.
        ...fragmentVariables,
        count,
        cursor,
      }
    },
    query: graphql`
      query WorksForYouQuery($count: Int!, $cursor: String) {
        ...WorksForYou_query @arguments(count: $count, cursor: $cursor)
      }
    `,
  }
)

export default WorksForYouContainer
