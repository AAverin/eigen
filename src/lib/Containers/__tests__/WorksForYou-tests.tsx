import React from "react"
import { NativeModules } from "react-native"
import * as renderer from "react-test-renderer"

import { renderWithLayout } from "../../tests/renderWithLayout"
import { WorksForYou } from "../WorksForYou"

import { Theme } from "@artsy/palette"

beforeAll(() => {
  NativeModules.ARTemporaryAPIModule = { markNotificationsRead: jest.fn() }
  WorksForYou.prototype.componentDidUpdate = () => {
    return null
  }
})

describe("with notifications", () => {
  it("creates a ListViewDataSource upon instantiation", () => {
    const worksForYou = new WorksForYou(notificationsResponse() as any)
    expect(worksForYou.state.dataSource).toBeTruthy()
  })

  it("updates the notification count", () => {
    const query = notificationsResponse().query
    renderer
      .create(
        <Theme>
          <WorksForYou query={query as any} relay={null} />
        </Theme>
      )
      .toJSON()
    expect(NativeModules.ARTemporaryAPIModule.markNotificationsRead).toBeCalled()
  })

  it("lays out correctly on small screens", () => {
    const query = notificationsResponse().query
    const component = renderWithLayout(<WorksForYou query={query as any} relay={null} />, { width: 100 })
    expect(component).toMatchSnapshot()
  })

  it("lays out correctly on larger screens", () => {
    const query = notificationsResponse().query
    const component = renderWithLayout(<WorksForYou query={query as any} relay={null} />, { width: 700 })
    expect(component).toMatchSnapshot()
  })
})

describe("when it has a special notification", () => {
  it("properly formats it and adds it to the top of the dataSource blob", () => {
    const response = selectedArtistResponse()
    const worksForYou = new WorksForYou(response as any)
    const expectedFormattedNotification = {
      id: "notification-juliana-huxtable",
      message: "1 Work Added",
      artists: "Juliana Huxtable",
      artworks: selectedArtistResponse().query.selectedArtist.artworks,
      image: {
        resized: {
          url: "cloudfront.url",
        },
      },
      artistHref: "artist/juliana-huxtable",
    }
    expect(worksForYou.state.dataSource.getRowData(0, 0)).toEqual(expectedFormattedNotification)
  })
})

describe("without notifications", () => {
  it("does not create a ListViewDataSource", () => {
    const worksForYou = new WorksForYou(emptyStateResponse() as any)
    expect(worksForYou.state.dataSource).toBeFalsy()
  })

  it("lays out correctly on small screens", () => {
    const query = emptyStateResponse().query
    const component = renderWithLayout(<WorksForYou query={query as any} relay={null} />, { width: 100 })
    expect(component).toMatchSnapshot()
  })

  it("lays out correctly on larger screens", () => {
    const query = emptyStateResponse().query
    const component = renderWithLayout(<WorksForYou query={query as any} relay={null} />, { width: 700 })
    expect(component).toMatchSnapshot()
  })
})

interface NotificationsResponse {
  query: {
    me: {
      followsAndSaves: {
        notifications: {
          pageInfo: {
            hasNextPage: boolean
            endCursor: string
          }
          edges: Array<{
            node: {
              artists: string
              summary: string
              artworks: { edges: [{ title: string }] }
              image: {
                resized: {
                  url: string
                }
              }
            }
          }>
        }
      }
    }

    selectedArtist?: any
  }
}

const notificationsResponse = () => {
  return {
    query: {
      me: {
        followsAndSaves: {
          notifications: {
            pageInfo: {
              hasNextPage: true,
              endCursor: null,
            },
            edges: [
              {
                node: {
                  artists: "Jean-Michel Basquiat",
                  summary: "1 Work Added",
                  artworks: { edges: [{ title: "Anti-Product Postcard" }] },
                  image: {
                    resized: {
                      url: "cloudfront.url",
                    },
                  },
                },
              },
              {
                node: {
                  artists: "Ana Mendieta",
                  summary: "2 Works Added",
                  artworks: { edges: [{ title: "CorazÃ³n de Roca con Sangre" }, { title: "Butterfly" }] },
                  image: {
                    resized: {
                      url: "cloudfront.url",
                    },
                  },
                },
              },
            ],
          },
        },
      },
    },
  } as NotificationsResponse
}

const emptyStateResponse = () => {
  return {
    query: {
      me: {
        followsAndSaves: {
          notifications: {
            pageInfo: {
              hasNextPage: true,
            },
            edges: [],
          },
        },
      },
    },
  }
}

const selectedArtistResponse = () => {
  {
    const response = notificationsResponse()
    response.query.selectedArtist = {
      slug: "juliana-huxtable",
      name: "Juliana Huxtable",
      href: "artist/juliana-huxtable",
      image: {
        resized: {
          url: "cloudfront.url",
        },
      },
      artworks: {
        edges: [
          {
            slug: "4594385943",
            title: "Untitled (Casual Power)",
          },
        ],
      },
    }
    return response
  }
}
