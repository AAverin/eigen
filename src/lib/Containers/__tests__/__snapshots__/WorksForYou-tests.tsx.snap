// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`with notifications lays out correctly on larger screens 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Text
    fontFamily="Unica77LL-Regular"
    fontSize="16px"
    lineHeight="26px"
    mb={1}
    mt={2}
    style={
      Array [
        Object {
          "fontFamily": "Unica77LL-Regular",
          "fontSize": 16,
          "lineHeight": 26,
          "marginBottom": 10,
          "marginTop": 20,
          "textAlign": "center",
        },
        Object {},
      ]
    }
    textAlign="center"
  >
    New Works for You
  </Text>
  <View
    style={
      Array [
        Object {
          "borderBottomWidth": 0,
          "borderColor": "#E5E5E5",
          "borderStyle": "solid",
          "borderWidth": 1,
          "width": "100%",
        },
      ]
    }
    width="100%"
  />
  <RCTScrollView
    ItemSeparatorComponent={[Function]}
    ListEmptyComponent={null}
    ListFooterComponent={null}
    data={Array []}
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onContentSizeChange={[Function]}
    onEndReached={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    refreshControl={
      <RefreshControlMock
        onRefresh={[Function]}
        refreshing={false}
      />
    }
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <RCTRefreshControl />
    <View />
  </RCTScrollView>
</View>
`;

exports[`with notifications lays out correctly on small screens 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Text
    fontFamily="Unica77LL-Regular"
    fontSize="16px"
    lineHeight="26px"
    mb={1}
    mt={2}
    style={
      Array [
        Object {
          "fontFamily": "Unica77LL-Regular",
          "fontSize": 16,
          "lineHeight": 26,
          "marginBottom": 10,
          "marginTop": 20,
          "textAlign": "center",
        },
        Object {},
      ]
    }
    textAlign="center"
  >
    New Works for You
  </Text>
  <View
    style={
      Array [
        Object {
          "borderBottomWidth": 0,
          "borderColor": "#E5E5E5",
          "borderStyle": "solid",
          "borderWidth": 1,
          "width": "100%",
        },
      ]
    }
    width="100%"
  />
  <RCTScrollView
    ItemSeparatorComponent={[Function]}
    ListEmptyComponent={null}
    ListFooterComponent={null}
    data={Array []}
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onContentSizeChange={[Function]}
    onEndReached={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    refreshControl={
      <RefreshControlMock
        onRefresh={[Function]}
        refreshing={false}
      />
    }
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <RCTRefreshControl />
    <View />
  </RCTScrollView>
</View>
`;

exports[`without notifications lays out correctly on larger screens 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Text
    fontFamily="Unica77LL-Regular"
    fontSize="16px"
    lineHeight="26px"
    mb={1}
    mt={2}
    style={
      Array [
        Object {
          "fontFamily": "Unica77LL-Regular",
          "fontSize": 16,
          "lineHeight": 26,
          "marginBottom": 10,
          "marginTop": 20,
          "textAlign": "center",
        },
        Object {},
      ]
    }
    textAlign="center"
  >
    New Works for You
  </Text>
  <View
    style={
      Array [
        Object {
          "borderBottomWidth": 0,
          "borderColor": "#E5E5E5",
          "borderStyle": "solid",
          "borderWidth": 1,
          "width": "100%",
        },
      ]
    }
    width="100%"
  />
  <RCTScrollView
    ItemSeparatorComponent={[Function]}
    ListEmptyComponent={null}
    ListFooterComponent={null}
    data={Array []}
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onContentSizeChange={[Function]}
    onEndReached={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    refreshControl={
      <RefreshControlMock
        onRefresh={[Function]}
        refreshing={false}
      />
    }
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <RCTRefreshControl />
    <View />
  </RCTScrollView>
</View>
`;

exports[`without notifications lays out correctly on small screens 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Text
    fontFamily="Unica77LL-Regular"
    fontSize="16px"
    lineHeight="26px"
    mb={1}
    mt={2}
    style={
      Array [
        Object {
          "fontFamily": "Unica77LL-Regular",
          "fontSize": 16,
          "lineHeight": 26,
          "marginBottom": 10,
          "marginTop": 20,
          "textAlign": "center",
        },
        Object {},
      ]
    }
    textAlign="center"
  >
    New Works for You
  </Text>
  <View
    style={
      Array [
        Object {
          "borderBottomWidth": 0,
          "borderColor": "#E5E5E5",
          "borderStyle": "solid",
          "borderWidth": 1,
          "width": "100%",
        },
      ]
    }
    width="100%"
  />
  <RCTScrollView
    ItemSeparatorComponent={[Function]}
    ListEmptyComponent={null}
    ListFooterComponent={null}
    data={Array []}
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onContentSizeChange={[Function]}
    onEndReached={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    refreshControl={
      <RefreshControlMock
        onRefresh={[Function]}
        refreshing={false}
      />
    }
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <RCTRefreshControl />
    <View />
  </RCTScrollView>
</View>
`;
